### ISSUE TITLE
Cannot push from shallow fetch

### ISSUE THREAD

**Issue by ckclark** [dulwich/dulwich#705]
## Environment

OS: linux
version: dulwich==0.19.11

## Steps to reproduce
```python
from dulwich.porcelain import push, clone, add, commit
repo = clone('<git-url>', 'out', depth=1)
fp = open('out/.gitignore', 'w')
fp.write('*.pyc')
fp.close()
add(repo, 'out/.gitignore')
commit(repo, b'Add .gitignore')
push(repo, '<git-url>', 'master')
```
This yields
```
Traceback (most recent call last):
  File "filename.py", line 197, in <module>
    push(repo, '<git-url>', 'master')
  File "/library/path/to/dulwich/porcelain.py", line 822, in push
    progress=errstream.write)
  File "/library/path/to/dulwich/client.py", line 784, in send_pack
    ofs_delta=(CAPABILITY_OFS_DELTA in negotiated_capabilities))
  File "/library/path/to/dulwich/object_store.py", line 248, in generate_pack_data
    self.generate_pack_contents(have, want, progress))
  File "/library/path/to/dulwich/object_store.py", line 236, in generate_pack_contents
    return self.iter_shas(self.find_missing_objects(have, want, progress))
  File "/library/path/to/dulwich/object_store.py", line 211, in find_missing_objects
    get_parents=get_parents)
  File "/library/path/to/dulwich/object_store.py", line 1110, in __init__
    have_commits, get_parents=self._get_parents)[0]
  File "/library/path/to/dulwich/object_store.py", line 288, in _collect_ancestors
    cmt = self[e]
  File "/library/path/to/dulwich/object_store.py", line 120, in __getitem__
    type_num, uncomp = self.get_raw(sha)
  File "/library/path/to/dulwich/object_store.py", line 482, in get_raw
    raise KeyError(hexsha)
KeyError: '<parent-commit-id>'
```
The object identified by parent-commit-id is not pulled due to the `depth=1` config. But this operation works if git cmdline is used.

```shell
$ git clone --depth=1 <git-url> out
Cloning into 'out'...
remote: Enumerating objects: 5, done.
remote: Counting objects: 100% (5/5), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 5 (delta 0), reused 4 (delta 0), pack-reused 0
Receiving objects: 100% (5/5), done.
$ cd out
$ echo "*.pyc" > .gitignore
$ git add .gitignore
$ git commit -m 'Add .gitignore'
[master 8fd5607] Add .gitignore
 1 file changed, 1 insertion(+)
 create mode 100644 .gitignore
$ git push origin master
Enumerating objects: 4, done.
Counting objects: 100% (4/4), done.
Delta compression using up to 72 threads
Compressing objects: 100% (2/2), done.
Writing objects: 100% (3/3), 316 bytes | 316.00 KiB/s, done.
Total 3 (delta 0), reused 0 (delta 0)
To <git-url>
   0ae984d..8fd5607  master -> master
```

**Comment by brechtm** [comment]
Any idea as to what the cause could be for this? I guess it needs to check whether the ancestors are present on the remote, not on the local repository.

**Comment by jelmer** [comment]
Fixed by @brechtm , will be in 0.19.17 and 0.20.0

**Comment by luckydonald** [comment]
To anyone confused, that fix is over at https://github.com/dulwich/dulwich/pull/757.
